linker/Linker.Steps/MarkStep.cs

```diff
@@ -468,9 +468,27 @@ protected virtual bool ShouldMarkCustomAttribute (CustomAttribute ca, ICustomAtt
 			return true;
 		}
 
-		protected virtual bool ShoulMarkTypeConstructor (TypeDefinition type)
+		protected virtual bool ShouldMarkTypeStaticConstructor (TypeDefinition type)
 		{
-			return !type.IsBeforeFieldInit && _context.IsOptimizationEnabled (CodeOptimizations.BeforeFieldInit);
+			if (Annotations.HasPreservedStaticCtor (type))
+				return false;
+			
+			if (type.IsBeforeFieldInit && _context.IsOptimizationEnabled (CodeOptimizations.BeforeFieldInit))
+				return false;
+
+			return true;
+		}
+
+		protected void MarkStaticConstructor (TypeDefinition type, bool force = false)
```

Comment:
Please keep the logic separated this should do only
```
				if (MarkMethodIf (type.Methods, IsStaticConstructor))
					Annotations.SetPreservedStaticCtor (type);
```

Reply:
Fixed